GitOut/Features/Git/Log/GitTreeEvent.cs

```diff
@@ -0,0 +1,154 @@
+                nodes.Add(node);
+                if (Event.Parent != null)
+                {
+                    bottomLeafs.Add(TreeBuildingLeaf.WithParent(Event.Parent, node));
+                    if (Event.MergedParent != null)
+                    {
+                        var mergedNode = GitTreeNode.WithBottomLine(new Line(from, to++), GetNextAvailableColor(), true);
+                        nodes.Add(mergedNode);
+                        bottomLeafs.Add(TreeBuildingLeaf.WithParent(Event.MergedParent, mergedNode));
+                    }
+                }
+            }
+            return bottomLeafs;
+        }
+
+        private Color GetNextAvailableColor()
+        {
+            AvailableColor nextColor = colors.FirstOrDefault(ac => ac.Available);
+            if (nextColor is null)
+            {
+                colors.ForEach(ac => ac.Available = true);
+                nextColor = colors[0];
+            }
+            nextColor.Available = false;
+            return nextColor.Color;
+        }
+
+        private void SetColorAvailable(Color color)
+        {
+            AvailableColor availableColor = colors.FirstOrDefault(ac => ac.Color == color);
+            if (availableColor != null)
+            {
+                availableColor.Available = true;
+            }
+        }
+
+        private IEnumerable<TreeBuildingLeaf> ProcessTop(IEnumerable<TreeBuildingLeaf> leafs)
+        {
+            int from = 0;
+            int to = 0;
+            var topLeafs = new List<TreeBuildingLeaf>();
+            foreach (TreeBuildingLeaf leaf in leafs)
+            {
+                if (leaf.LookingFor == Event)
+                {
+                    GitTreeNode node;
+                    if (CommitIndex == -1)
+                    {
+                        CommitIndex = to++;
+                        node = GitTreeNode.WithTopLine(new Line(from++, CommitIndex), leaf.Current.Color, true);
+
+                        // add node to topleafs, so that it's processed in ProcessBottom
+                        // done only for the first branch that finds the commit, the other would be duplicates
+                        // the LookingFor is unused later - it is the parents of the Event that will be used as LookingFor
+                        topLeafs.Add(TreeBuildingLeaf.WithoutParent(node));
+                    }
+                    else
+                    {
+                        node = GitTreeNode.WithTopLine(new Line(from++, CommitIndex), leaf.Current.Color, true);
+                        SetColorAvailable(leaf.Current.Color);
+                    }
+                    nodes.Add(node);
+                }
+                else
+                {
+                    var newNode = GitTreeNode.WithTopLine(new Line(from++, to++), leaf.Current.Color, false);
+                    nodes.Add(newNode);
+                    topLeafs.Add(TreeBuildingLeaf.WithParent(leaf.LookingFor!, newNode));
```

Comment:
Instead of using the null-forgiving operator we could also add a null check and throw an InvalidOperationException. What do you think?
```suggestion
                    if (leaf.LookingFor is null) { throw new InvalidOperationException("Parent should not be null"); }
                    topLeafs.Add(TreeBuildingLeaf.WithParent(leaf.LookingFor, newNode));
```